### DISCUSSION TITLE
For loops in nn.Module

### DISCUSSION THREAD

**Discussion by AkashGanesan** [google/flax#1639] [Q&A] [answered]
We have a top level module that has several submodules that can be executed in parallel. In the following code, we loop over the branches. However, we are afraid that these won't be optimized by jit and will execute sequentially. Can you confirm if this is fine or if it, provide an example on how to actually achieve this?

```python
import jax
import jax.numpy as jnp


class Model(nn.Module):

    branches: dict

    def setup(self):
        branch = []
        for branch_name, branch_params in branches.items():
            b = branch_params.type_(**branch_params.params)
            branch.append(b)
        self.branch = branch

    def __call__(self, x):
        """
        Forward pass branches.

        Args:
            x(tuple(jnp.array)): Shape is (batch_size, width, height, channels, num_branches)
        """
        outs = []

        for idx, b in enumerate(self.branch):
            outs.append(b(x[..., idx]))

        return outs

**Comment by jheek** [comment, marked as answer]
This depends very much on the device and optimizations made by XLA. Mostly XLA uses parallelization within a single op rather than executing multiple ops in parallel. If branches are similar enough you could potentially vectorize over them with `vmap` but this will not work if the branches are actually doing different things (executing different ops). I think the best things here is to avoid falling for the trap of premature optimization. So let XLA do it's magic unless you have clear evidence that it's doing a poor job (for example low CPU/GPU/TPU utilization)

**Reply by AkashGanesan** [reply]
Issue is, we are getting low GPU utilization (~ 1/3).   Is there a way to look at the code generated by XLA?  Also, this is to make a framework for self supervised learning and this is with  BYOL (we are testing it on CIFAR10).  For an input dimension of 32x32x3 and a batch size of 128, it takes around 30 seconds.  I single resnet50 pass is less than a second for the same size so I'd think it'd be around the same order.

**Reply by AkashGanesan** [reply]
The utilization is way higher without batchnorm.  I removed all BNs from the resnet50 and the forward pass is under a second now (as opposed to 30s).

**Reply by AkashGanesan** [reply]
We found that this comes from `state.apply_gradients` (Followed the imagenet example on extending Trainstate to include batch_stats)

I tried the following

```python

## This take ~ 30s
state = state.apply_gradients(
                grads=grad["params"],
                batch_stats=batch_stats

## This take around ~ 0.04 seconds
state = state.apply_gradients(
                grads=grad["params"],)                

## This takes  around ~30s
state = state.apply_gradients(
                grads=grad["params"],
                batch_stats=aux
            )
state = state.replace(batch_stats=aux)
```

It's probably moving something off the GPU here?

**Reply by AkashGanesan** [reply]
I'll create a new discussion as this is no longer appropriate.  Marking the answer as accepted.
